vocabs/v1/context.jsonld

```diff
@@ -1,90 +1,61 @@
-    "Bls12381G1Key2020": "sec:Bls12381G1Key2020",
-    "Bls12381G2Key2020": "sec:Bls12381G2Key2020",
-    "RsaVerificationKey2018": "sec:RsaVerificationKey2018",
-    "SchnorrSecp256k1VerificationKey2019": "sec:SchnorrSecp256k1VerificationKey2019",
-    "X25519KeyAgreementKey2019": "sec:X25519KeyAgreementKey2019",
-    "ServiceEndpointProxyService": "didns:ServiceEndpointProxyService",
-    "LinkedDomains": "https://identity.foundation/.well-known/resources/did-configuration/#LinkedDomains",
+
     "alsoKnownAs": {
       "@id": "https://www.w3.org/ns/activitystreams#alsoKnownAs",
       "@type": "@id",
       "@container": "@set"
     },
     "assertionMethod": {
-      "@id": "sec:assertionMethod",
+      "@id": "https://w3id.org/security#assertionMethod",
       "@type": "@id",
       "@container": "@set"
     },
     "authentication": {
-      "@id": "sec:authenticationMethod",
+      "@id": "https://w3id.org/security#authenticationMethod",
       "@type": "@id",
       "@container": "@set"
     },
     "capabilityDelegation": {
-      "@id": "sec:capabilityDelegationMethod",
+      "@id": "https://w3id.org/security#capabilityDelegationMethod",
       "@type": "@id",
       "@container": "@set"
     },
     "capabilityInvocation": {
-      "@id": "sec:capabilityInvocationMethod",
+      "@id": "https://w3id.org/security#capabilityInvocationMethod",
       "@type": "@id",
       "@container": "@set"
     },
     "controller": {
-      "@id": "sec:controller",
+      "@id": "https://w3id.org/security#controller",
       "@type": "@id"
     },
-    "created": {
-      "@id": "dc:created",
-      "@type": "xsd:dateTime"
-    },
-    "blockchainAccountId": "sec:blockchainAccountId",
     "keyAgreement": {
-      "@id": "sec:keyAgreementMethod",
-      "@type": "@id",
-      "@container": "@set"
-    },
-    "publicKey": {
-      "@id": "sec:publicKey",
+      "@id": "https://w3id.org/security#keyAgreementMethod",
       "@type": "@id",
       "@container": "@set"
     },
-    "publicKeyBase58": "sec:publicKeyBase58",
-    "publicKeyJwk": {
-      "@id": "sec:publicKeyJwk",
-      "@type": "@json"
-    },
     "service": {
-      "@id": "didns:service",
-      "@type": "@id",
-      "@container": "@set"
-    },
-    "serviceEndpoint": {
-      "@id": "didns:serviceEndpoint",
-      "@type": "@id"
-    },
-    "updated": {
-      "@id": "dc:modified",
-      "@type": "xsd:dateTime"
+      "@id": "https://www.w3.org/ns/did#service",
+      "@type": "@id",
+      "@container": "@set",
+      "@context": {
+        "@protected": true,
+        "id": "@id",
+        "type": "@type",
+        "serviceEndpoint": {
+          "@id": "https://www.w3.org/ns/did#serviceEndpoint",
+          "@type": "@id"
+        }
+      }
     },
     "verificationMethod": {
-      "@id": "sec:verificationMethod",
-      "@type": "@id"
+      "@id": "https://w3id.org/security#verificationMethod",
+      "@type": "@id",
+      "@container": "@set"
```

Comment:
I agree that this should not be a `@set` -- it wasn't in an earlier revision of this PR.

```suggestion
      "@type": "@id"
```

Reply:
assuming we accept this change, won't this context allow you to embed an object? seems like thats actually incompatible with the did core spec..... I have never once seen a did document that allowed a verification relationship to be anything but an array... and I object to implying that its a good idea outside of the actual spec.

Reply:
strong -1 to did context allowing non conformant data structures... feel free to do that in another context.... not sure I am reading this correctly, but it sounds like you are saying the did core context should be more flexible than the did core spec... I don't agree.

Reply:
@OR13, No, I'm saying that the did core context does not work the way you seem to be implying. No context does. It's not a schema.

Reply:
@dlongley is right, schema enforcement is done at another layer (JSON Schema, or hand-written schema enforcement rules). 

@dlongley, I will note that this raises an entirely separate issue... all the verification relationships have `"@container": "@set"` for the reason @OR13 states above. So, now the context is not internally consistent... why are we specifying that verification relationships are sets and `verificationMethod` is not?

Remember that we had these rules in DID Core because people didn't want to guess if something was an array or not in a DID Document (and we're using JSON-LD to try and force everything into an array by default). I believe using JSON-LD in this way is probably misguided, resolver libraries are perfectly capable of putting the DID Document into a canonical "everything is an array" form for developers... but here we are. Note that the JSON-only processors WILL NOT have this safety unless folks put it in there, so one potential path forward is to NOT do `@container: @set` for all these properties and push the burden of putting these things into array form to the resolver libraries. 

Thoughts?

Reply:
> So, now the context is not internally consistent... why are we specifying that verification relationships are sets and `verificationMethod` is not?

Because, for better or worse, `verificationMethod` plays multiple roles. It is used in proofs to reference the verification method used (where it refers to a single method) and used as a catch-all bucket for all verification methods (where it refers to multiple methods). The term is protected and we don't want to deal with clashes in the ecosystem so we need to pick one preferred representation. Since:

1. it is not best security practice to reuse verification relationships for many different purposes or for undefined purposes (which is what using the catch-all `verificationMethod` as a verification relationship is for), and
2. usage of it as a verification relationship not common, and
2. the usage of `verificationMethod` in proofs is already quite common place in the ecosystem, and
3. JSON-LD contexts are not schemas anyway,

it seemed most prudent to favor not marking it with `@set`. The other purpose-specific verification relationships have no similar issue.

Reply:
Mechanically, IF the context can enforce or describe that `verificationMethod` is always an array, it should. 

We should NEVER be implying that JSON-LD and JSON might have have different types for `verificationMethod`.

Since did core spec would need to be changed to allow a verification method to be an object, we should either make that change too did core, or we should make it as clear as humanly possible that verificationMethod and relationships are always arrays.

Reply:
> Mechanically, IF the context can enforce or describe that verificationMethod is always an array, it should.

It can't.

> Since did core spec would need to be changed to allow a verification method to be an object, we should either make that change too did core, or we should make it as clear as humanly possible that verificationMethod and relationships are always arrays.

I think the spec does that already here: https://www.w3.org/TR/did-core/#did-document-properties

The term is already defined across the ecosystem without the `@set` preference. If that were not the case, I'd be onboard with using `@set` because there would be no reason to do otherwise. However, given the reality, I think requiring the context to define the term in a different way from its existing use in the ecosystem would lead to needless incompatibilities and frustration. Defining it the same way, alternatively, may only require a small tweak to some resolvers to make sure an array is produced in their output. This is the better outcome, IMO.

Reply:
@OR13 -- just to be crystal clear. If we made this `@container`: `@set`, it would absolutely guarantee that no DID Document could ever be digitally signed... you know, like `did:web`... no signatures on `did:web`... ever. Not even self-signed documents. I don't think you want that.

This is because `verificationMethod` is not only used in DID Documents, but it's used in Linked Data Signatures as well. In the DID JSON-LD Context, IF we were to define it like so:

```
   "@protected": "true",
   ...
    "verificationMethod": {
      "@id": "https://w3id.org/security#verificationMethod",
      "@type": "@id",
      "@container": "@set" <--- this is a landmine, see below
    }
```

and then, in the security vocab (or any cryptosuite vocab) it was defined as so:

```
   "@protected": "true",
   ...
    "verificationMethod": {
      "@id": "https://w3id.org/security#verificationMethod",
      "@type": "@id" <--- note, "@container": "@set" is missing
    }
```

Then you will get a protected term redefinition error for `verificationMethod`.

So, something has to give. Either we always make `verificationMethod` an array in `proof`s (which makes no sense, you can't have two public keys be the verification method for a single digital signature), or we make it possible for a verification method to be either an array of values or a single value. The latter enables all the use cases we know of.

The place to enforce this is in DID Document output... we can keep the text that's in the spec now and it's the job of a JSON-LD producer to always ensure that the verificationMethod is associated with an array. It won't happen automatically, like it does for the verification relationships, but it is possible to make sure that there is no difference between JSON output and JSON-LD output.

In hindsight, it was a mistake to enforce this at the JSON-LD processor layer... it should've been enforced at the DID Document JSON-LD production layer. The end result is the same and gets @OR13 what he wants.

Thoughts, @OR13?

Reply:
> which makes no sense, you can't have two public keys be the verification method for a single digital signature

https://github.com/w3c-ccg/community/issues/188

Also aggregate signature schemes generally.

We should make did core the same as security vocab.

https://github.com/w3c-ccg/security-vocab/blob/master/contexts/security-v2.jsonld#L55

> It won't happen automatically, like it does for the verification relationships, but it is possible to make sure that there is no difference between JSON output and JSON-LD output.

Perhaps we should also remove `@container`: `@set` from them as well...

https://github.com/w3c/did-spec-registries/blob/main/vocabs/v1/context.jsonld#L60

Consistency is worth more here IMO, if we are in a world where you need to make sure 1 thing is an array, why not make it so that all relationships are treated the same.

Reply:
> Also aggregate signature schemes generally.

Yeah, we need to discuss that as a community. I have my concerns based on some proposed structures (and `verificationMethod` ever being allowed to be an array in a proof -- there may be use cases there, but one use case we definitely have today is the "not an array" use case): 

https://github.com/w3c-ccg/community/issues/188#issuecomment-812887907

> We should make did core the same as security vocab.

Agreed.

I thought you were arguing against this, seems like you aren't anymore?

> Consistency is worth more here IMO, if we are in a world where you need to make sure 1 thing is an array, why not make it so that all relationships are treated the same.

I'm fine w/ doing that. That would mean none of those would be associated with "@container": "@set", and instead, JSON-LD Producers have to ensure that they're sets when they produce the DID Document (which is the normative requirement in the spec today).

Reply:
Please be aware that *today* all of the verification relationships that aren't the "catch all" `verificationMethod` are defined as `@set`s in contexts across the ecosystem. Deciding to make those now NOT `@set`s in the DID context "just for consistency" is ... not consistent. This latter inconsistency is more negatively impactful to the existing ecosystem.

Reply:
> Deciding to make those now NOT @sets in the DID context "just for consistency"

It's not being done for consistency... I'm starting to think that `"@container": "@set"` is an anti-pattern when used with protected contexts... it creates more problems than it's worth.

Reply:
>  it creates more problems than it's worth.

I agree with that.

@msporny I adjusted my argument here:

https://github.com/w3c/did-spec-registries/pull/277#pullrequestreview-627464044

Essentially DID Core should go all in on sec-v2 or all in on sec-v3 (really this is sec-v3 following did core instead of sec-v2 on this issue).

I don't thing not breaking LD Signatures on a DID Document is a compelling reason to avoid consistency on this matter.

sec-v2 cannot be changed, did core and sec-v3 can... whatever decision we make should be present in both.

Reply:
Reminder that the VC context uses `@set` in its proof purpose definitions -- so that should also be considered as a potential hazard.

I'm starting to think that a future revision to JSON-LD's protected term feature should consider allowing `@set` to be ignored since it doesn't change the meaning of the term, only the *preferred* shape when compaction is used -- which doesn't really matter when you're intentionally not compacting anything and simply validating your inputs using JSON schema, etc. (@iherman, this is something to consider in the future / or perhaps an "evergreen" thing).